crates/codegen/parser/runtime/src/cst.rs

```diff
@@ -46,63 +46,103 @@ impl Node {
         }
     }
 
-    pub fn create_cursor(&self, text_offset: TextIndex) -> Cursor {
+    /// Creates a [`Cursor`] that starts at the current node as the root and a given initial `text_offset`.
+    pub fn cursor_with_offset(&self, text_offset: TextIndex) -> Cursor {
         Cursor::new(self.clone(), text_offset)
     }
 
+    /// Reconstructs the original source code from the parse tree.
+    pub fn reconstruct(self) -> String {
+        match self {
+            Self::Rule(rule) => rule.reconstruct(),
+            Self::Token(token) => token.text.clone(),
+        }
+    }
+
     pub fn as_rule(&self) -> Option<&Rc<RuleNode>> {
         match self {
             Self::Rule(node) => Some(node),
             _ => None,
         }
     }
 
+    pub fn into_rule(self) -> Option<Rc<RuleNode>> {
+        match self {
+            Self::Rule(node) => Some(node),
+            _ => None,
+        }
+    }
+
     pub fn as_token(&self) -> Option<&Rc<TokenNode>> {
         match self {
             Self::Token(node) => Some(node),
             _ => None,
         }
     }
 
-    pub fn as_token_with_kind(&self, kinds: &[TokenKind]) -> Option<&Rc<TokenNode>> {
-        if let Node::Token(token_node) = self {
-            if kinds.contains(&token_node.kind) {
-                return Some(token_node);
-            }
+    pub fn into_token(self) -> Option<Rc<TokenNode>> {
+        match self {
+            Self::Token(node) => Some(node),
+            _ => None,
         }
-        return None;
+    }
+
+    pub fn as_token_with_kind(&self, kinds: &[TokenKind]) -> Option<&Rc<TokenNode>> {
+        self.as_token_matching(|token| kinds.contains(&token.kind))
     }
 
     pub fn as_token_matching<F: Fn(&Rc<TokenNode>) -> bool>(
         &self,
         predicate: F,
     ) -> Option<&Rc<TokenNode>> {
-        if let Node::Token(token_node) = self {
-            if predicate(&token_node) {
-                return Some(token_node);
-            }
+        match self {
+            Node::Token(token) if predicate(&token) => Some(token),
+            _ => None,
         }
-        return None;
     }
 
     pub fn as_rule_with_kind(&self, kinds: &[RuleKind]) -> Option<&Rc<RuleNode>> {
-        if let Node::Rule(rule_node) = self {
-            if kinds.contains(&rule_node.kind) {
-                return Some(rule_node);
-            }
-        }
-        return None;
+        self.as_rule_matching(|rule| kinds.contains(&rule.kind))
     }
 
     pub fn as_rule_matching<F: Fn(&Rc<RuleNode>) -> bool>(
         &self,
         predicate: F,
     ) -> Option<&Rc<RuleNode>> {
-        if let Node::Rule(rule_node) = self {
-            if predicate(&rule_node) {
-                return Some(rule_node);
-            }
+        match self {
+            Node::Rule(rule) if predicate(&rule) => Some(rule),
+            _ => None,
         }
-        return None;
+    }
+}
+
+impl From<Rc<RuleNode>> for Node {
```

Comment:
Possibly unrelated to this PR, but asking since it is addressing API feedback:

I like using `std::` traits like `impl From<Rc<RuleNode>> for Node`.
But we only do that here, and in the other way around, we use `as_token()` and `as_rule()`. I wonder if instead of these two, we should `impl TryFrom<Node>` for `Token` and `Rule` instead?

Reply:
I added `From` for completeness' sake and for others to maybe conveniently consume these argument in the `Into<Node>` type position, but we should still be using the regular enum constructors internally IMO, as it's more readable (e.g. `rule.into()` with bigger cognitive load vs `Node::Rule(rule)`).

As for `TryFrom`/`TryInto`, these traits are great on paper and they're finally included in the Rust 2021 prelude (it was one of the biggest paper cuts) but the fact that:
1. it returns a `Result` - it's not really an "error"; `std` implementations use this to validate the type (bit, array) size or invariants such as UTF-8 in `OsStr`->`str` conversion. Working around that with a `Result<Node, ()>` makes this a bit unwieldy and doesn't fit the use case quite well
2. We can benefit from the ergonomic, inferred `.try_from/into()` only in a place where the type is already given/expected, however we mostly prefer to unwrap it ourselves at call-site as a given type and...
3. ...doing so using the traits would require us to use the fully qualified syntax, like so: `TryInto::<Rc<Rule>>::try_into(node)` or `<Rc<Rule> as TryFrom<Node>>::try_from(node)` (versus `node.as_rule()`).

As such, I think it's more idiomatic and simpler to provide only the explicit conversion `as/into` methods as there's not really much value in implementing this trait in this use case, either for us or for the public API consumers in my opinion.

More about the ad-hoc conversion and conventions: https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv

Reply:
> Working around that with a `Result<Node, ()>` makes this a bit unwieldy and doesn't fit the use case quite well

But it can be an error? if you try to convert a node (that contains a token) into a rule node, it should fail, right?

Reply:
Only in a very theoretical sense, yes.

However, the users will not care or handle the possible error value (`()` or a theoretical unit struct `NotToken`), nor will they benefit from `Result`s combinators. We will also not benefit from the implementation nor will the users, as application of the trait is limited in this case, as per above.

I have given an in-depth reasoning above, with a link for further context on conventions for trivial conversions such as this, and I feel like I'm not really heard here.

Reply:
Makes sense. Thanks for explaining!